laws/src/main/scala/cats/laws/SelectiveLaws.scala

```diff
@@ -0,0 +1,48 @@
+package cats
+package laws
+
+import cats.syntax.all._
+
+/**
+ * Laws that must be obeyed by any `Selective`.
+ */
+trait SelectiveLaws[F[_]] extends ApplicativeLaws[F] {
```

Comment:
Why if we make RigidSelectiveLaws that has this plus rigidity and MonadLaws extend that?

Reply:
"Why if" == "what if"?  In other words, you're proposing `RigidSelectiveLaws` between `SelectiveLaws` and `ApplicativeLaws`? I'm good with that.  

Does that imply an empty `RigidSelective` typeclass as a marker?  I'm not sure where we'd use it as a constraint, but it's weird to have laws with no corresponding typeclass.

Reply:
We'd need to express the rigidity law without `FlatMap`, or else there isn't much point.

Reply:
I think the law for parsers you want is `select(pa.map(Right))(pb) == pa`, so you don't want any action on the second parser if the first one succeeds and returns Right.

To test this with for instance IO, if you have some Ref, and the second item increments, if the first IO returns right, the increment should not happen.

In any case, I think the laws I want are:
```scala
select(pa.map(Right))(pb) == pa
select(pa.map(Left))(pb) == (pa, pb).mapN { (b, fn) => fn(b) }
```

Is that overdetermined?

Reply:
The Haskell implementation proposes two similar properties, true whenever `apS == <$>`:

```haskell
select (pure (Left  x)) y = ($x) <$> y -- Pure-Left
select (pure (Right x)) y = pure x     -- Pure-Right
```

But as long as `Eq[IO[_]]` is based on `unsafeRunSync()`, then I don't see how that law precludes evaluating `pb`.  I think we have to go to SyncLaws to generically test that the effect is not being evaluated.

The laws for rigid selectives in the paper are:

```haskell
-- Laws for rigid selective functors (in particular, for monads)
(<*>) = apS -- Selective apply
x *> (y <*? z) = (x *> y) <*? z -- Interchange
```

Reply:
I suppose we could also use tricks like the defer laws to ensure the optional effect isn't evaluated, and not go all the way to `Sync`.

Reply:
the law I suggest implies your law, and it can detect evaluating pb. Is it too strong of a law? I think it may be exactly the rigidity law stated a different way.

Reply:
Let's make these three laws "rigid".  One from the paper, and two from you:

```scala
  def selectiveApply[A, B](ff: F[A => B], fa: F[A]): IsEq[F[B]] =
    ff.apS(fa) <-> ff.ap(fa)

  def oscarsRight[A, B](fb: F[B], ff: F[A => B]): IsEq[F[B]] =
    fb.map(Right(_)).select(ff) <-> fb

  def oscarsLeft[A, B](fa: F[A], ff: F[A => B]): IsEq[F[B]] =
    fa.map(Left(_)).select(ff) <-> (fa, ff).mapN { (a, fn) => fn(a) }
```

Override `Select[Eval]`:

```scala
      override def select[A, B](fa: Eval[Either[A, B]])(ff: Eval[A => B]): Eval[B] =
        fa.flatMap {
          case Left(a) => ff.map(_(a))
          case Right(b) => ff.map(_ => b)
        }
```

It's lawful:

```
  + Eval[Int]: rigidSelective.rigid selective Oscar's left 0.001s
  + Eval[Int]: rigidSelective.rigid selective Oscar's right 0.0s
  + Eval[Int]: rigidSelective.rigid selective apply 0.001s
  + Eval[Int]: rigidSelective.selective associativity 0.001s
  + Eval[Int]: rigidSelective.selective branch consistency 0.001s
  + Eval[Int]: rigidSelective.selective distributivity 0.001s
  + Eval[Int]: rigidSelective.selective identity 0.001s
  + Eval[Int]: rigidSelective.selective ifS consistency 0.001s
  + Eval[Int]: rigidSelective.selective whenS consistency 0.001s
  + Eval[Int]: rigidSelective.semigroupal associativity 0.001s
```

But it evaluates `pb` even on Right:

```scala
scala> val pa = Eval.now("Hi")
val pa: cats.Eval[String] = Now(Hi)

scala> val pb = Eval.later(println("ope!")) *> Eval.now((i: Int) => i.toString)
val pb: cats.Eval[Int => String] = cats.Eval$$anon$4@370b3acf

scala> pa.map(Right(_)).select(pb).value === pa
ope!
val res14: Boolean = true
```

We can check anything in the purview of the the `Eq` instance, but I don't think we can legislate the evaluation of the second effect without either a `Defer` constraint (`Sync` unites `Defer` with `Select`), or by making the second argument lazy.

Reply:
yeah, it passes the law but it evaluates the right, but you had to make a side-effecting Eval to do that. I say you broke the rules at that point and you got what you asked for.

With IO, which allows you to put side effects in the IO, a proper equality includes all the state of the world that might change after the IO is run.

We could actually do that with Eval, and detect the bug here: make an Eval that reads from a mutable variable using Eval.always. With that, your law won't pass I believe.

btw: isn.t apS == ap just oscarsLeft in the above?

Reply:
If we don't have the Right law, I don't know how you really program against this typeclass: I may or may not evaluate both arguments even if I don't need to (judged by the types)... okay, I guess...

To me, without that law we are back to allowing Applicative as a valid Selective and I'm pretty firmly in the no-camp on superclasses being valid implementations of their subclasses. In fact: I would propose that as a principle in general: we should never admit a typeclass whose laws are so weak that superclasses can satifsy them.

Reply:
> With IO, which allows you to put side effects in the IO, a proper equality includes all the state of the world that might change after the IO is run.

Yes, `IO` makes a better example than `Eval`.  Like `Eq[Eval[A]]`, we don't have an `Eq[IO[A]]` that can detect the unwanted evaluation of `ff`.  (It should, but it's hard.)  Unlike `Eval`, side effects are welcome in `IO`.  I see two paths forward:

1. The law is that `ff` mustn't be evaluated for any effect type.  In that case, we should make `ff` by name.

2. The law is that `ff` mustn't perform unnecessary side effects.  In that case, we add a `SyncLaw`, which would cover `IO` and similar, and declare that my `Eval` example voids the warranty.

Since we're concerned about this property in `Parser`, which has no `Sync`, I'm leaning toward the former.

> btw: isn.t apS == ap just oscarsLeft in the above?

I haven't done the algebra yet, but the very similar `Pure-Left` and `Pure-Right` are proven in Coq. (I don't read Coq).

> If we don't have the Right law, I don't know how you really program against this typeclass: I may or may not evaluate both arguments even if I don't need to (judged by the types)... okay, I guess...

This flexibility is crucial to `Over` and `Under`, but that could be done with an `Applicative` constraint if `select` were pushed down.  

We might rearrange the laws like this:
* Selective associativity in `Apply`... I think.
* Distributivity and identity in `Applicative`.
* No laws, and no typeclass, for `Selective`.
* Rigidity by a new `RigidSelective` type class.  Few of these won't also be monads.
* selectM consistency as a Monad law.  I'm not sure it's interesting as long as all monads are rigid.
* Non-evaluation of second argument in RigidSelective if it's lazy, or Sync if it's not.

Reply:
I'm fairly unconvinced that Over and Under are the main use cases here.

I don't really love the name RigidSelective personally. I would just use Selective, personally, since there is no compelling other definition for Selective (disregarding the idea of making a new typeclass with methods that can be lawfully implemented by Apply/Applicative).

If you really want to put select on Apply, I would suggest we call it selectA, and add a law in (RigidSelective) that selectA == select.

I think the non-evaluation law I want is just the Right law (pa.map(Right).select(f) == pa). For a parser, you would definitely detect if select triggered f. I'm not really concerned with the Eval example since Eval is purely to model how a computation is done, so, presumably we wouldn't make a bad implementation, so it is fine.

I hear you that IO may or may not generate complex enough values in its Arbitraries (I think the work Erik showed which we developed at Stripe and I used in cats-parse could be used to make sure IO really exercises the possible space).

Lastly, I would reiterate my point about using a call-by-name for ff just to allow to implement ifM on (RigidSelective). I kind of hate introducing a new method that is identical to something on a subclass.

Finally, I think untilDefinedM can be defined with selective:

```
def untilDefinedM[A](fa: F[Option[A]]): F[A] = {
  val init: F[Either[Unit, A]] = fa.map {
    case Some(a) => Right(a)
    case None => Left(()) // once again an Either unit value
  }
  lazy val fn: F[Unit => A] = untilDefinedM(fa).map { a => (u: Unit) => a }
 
  // if the second param is call-by-name, we only loop as many times as needed
  select(init)(fn)
}
``` 
this an be an override in Monad which can use tailRecM which might be needed for very long chains.

Reply:
> If you really want to put select on Apply, I would suggest we call it selectA, and add a law in (RigidSelective) that selectA == select.

`selectA` seems reasonable, but I'd expect it to be strict in both arguments, like other `A` methods?  And then we still need `select` with a lazy second argument somewhere below RigidSelective if we want the `Over`/`Under` that you don't seem to want. :laughing: 

`select == selectM` is the law in Haskell once you get to Monad.  I don't think `select == selectA` and `select == selectM` are compatible as laws.